global.go

```diff
@@ -0,0 +1,248 @@
+// evaluations and no sense of local convergence
+type GlobalMethod interface {
+	// Need to tell the optimizer how many can be done at once (for example random restart)
+	InitGlobal(tasks int) // can this be concurrent bool? Seems like optimizer should be told how many tasks
+	// Global method may assume that the same task id always has the same pointer with it.
+	IterateGlobal(task int, loc *Location) (Operation, error)
+	Needs() struct{ Gradient, Hessian bool }
+}
+
+// Global uses a global optimizer to search for the gloabl minimum of a
+// function. A maximization problem can be transformed into a
+// minimization problem by multiplying the function by -1.
+//
+// Global, unlike Local, searches for the global minimum of the function.
+// GlobalMethods typically do not make any assumptions about the number and location
+// of local minima, and so the only convergence metric used is the convergence
+// of function values found at major iterations. GradientThreshold is not checked
+// during convergence, and initial values are not used.
+func Global(p Problem, dim int, settings *Settings, method GlobalMethod) (*Result, error) {
+	if method == nil {
+		// TODO(btracey): Replace with a better global minimizer (basically
+		// anything else). Though technically this isn't worse on the space of
+		// all problems.
+		method = &GuessAndCheck{}
+	}
+	if settings == nil {
+		settings = DefaultSettingsGlobal()
+	}
+	startTime, err := initializeOptimization(p, settings, method)
+	if err != nil {
+		return nil, err
+	}
+	optLoc := newLocation(dim, method)
+	optLoc.F = math.Inf(1)
+
+	gs := &globalStatus{
+		mux:       &sync.Mutex{},
+		stats:     &Stats{},
+		status:    NotTerminated,
+		p:         p,
+		startTime: startTime,
+		optLoc:    optLoc,
+		settings:  settings,
+	}
+
+	nTasks := settings.Concurrent
+	method.InitGlobal(nTasks)
+
+	// Launch optimization workers
+	var wg sync.WaitGroup
+	for task := 0; task < nTasks; task++ {
+		wg.Add(1)
+		go func(task int) {
+			defer wg.Done()
+			loc := newLocation(dim, method)
+			x := make([]float64, dim)
+			globalWorker(task, method, gs, loc, x)
+		}(task)
+	}
+	wg.Wait()
+
+	// Cleanup and collect results
+	if settings.Recorder != nil && err == nil {
+		// Send the optimal location to Recorder.
+		err = settings.Recorder.Record(gs.optLoc, PostIteration, gs.stats)
+	}
+	gs.stats.Runtime = time.Since(startTime)
+	return &Result{
+		Location: *gs.optLoc,
+		Stats:    *gs.stats,
+		Status:   gs.status,
+	}, err
+}
+
+type globalStatus struct {
+	mux       *sync.Mutex
+	stats     *Stats
+	status    Status
+	p         Problem
+	startTime time.Time
+	optLoc    *Location
+	settings  *Settings
+}
+
+func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []float64) {
+	for {
+		// Find Evaluation location
+		op, err := m.IterateGlobal(task, loc)
+		if err != nil {
+			// TODO(btracey): Figure out how to handle errors properly. Shut
```

Comment:
Whoa. Did not know that was a thing.

Barring that, I think it should probably shut down the process. The method is in charge of returning the error, so if it wants to handle an error internally it can.